pkg/apis/agones/v1/fleet.go

```diff
@@ -190,6 +190,9 @@ func (f *Fleet) Validate() ([]metav1.StatusCause, bool) {
 	if len(gsCauses) > 0 {
 		causes = append(causes, gsCauses...)
 	}
+	if productCauses := apiHooks.ValidateScheduling(f.Spec.Scheduling); len(productCauses) > 0 {
```

Comment:
Should these have new unit tests? Or do unit tests only make sense in the (existing) cloudproduct package where the behavior can return an error?

Reply:
My feeling was that the logic here was small enough that unit testing in the cloud product was sufficient. It was also perhaps expediency for the fact that there is not yet a fake cloud product for unit tests. If you'd like, I can address that here.

Reply:
+1 on having a unit test. Otherwise, this looks good to me 👍🏻

Reply:
I'm ok with unit tests in the provider package as long as we are still testing that this check is working properly in a unit test.

Reply:
The gke provider code path is exercised in `TestValidateGameServer`, which calls it through `ValidateGameServerSpec` -> `ValidateScheduling`. It's slightly different than the path at the cursor here, _but_ it's definitely exercised.

I looked at exercising `apiHooks` and I will say, it's a convenient structure and I still think it's better than the way I was previously handling it, but explicitly overriding `apiHooks` for tests is difficult: Unlike the feature flags, the apiHooks are called _broadly_ due to their presence in basic structure validation. I headed down a part of adding a mutex just like `runtime` handles its singleton, but it ended being fairly invasive and I'm still fighting. I'm looking at it and .. don't think it's the right approach.

Rather, I'm wondering if for both feature flags and cloud product hooks if we should instead be passing through some sort of context structure, which is the more Go-endorsed solution. It's equally invasive but it's rhythmic in that for the most part such a context would just flow through most functions, similar to passing `ctx` around.

But, short version: I don't think it's worth testing this exact line, and I think the GKE provider implementation is tested.

Reply:
I wouldn't mind seeing a test that was basically a "do causes get appended if there is a result from `apiHooks.ValidateScheduling`" just to confirm the branch works as expected.

I don't think it needs to be anything more complex than that.

Reply:
We compromised on an e2e test, which I added (by modifying one of the Skip cases).